opal/mca/btl/iwarp/btl_iwarp_component.c

```diff
@@ -0,0 +1,3788 @@
+    if (NULL == target->receive_queues && NULL != src->receive_queues) {
+        target->receive_queues = strdup(src->receive_queues);
+    }
+
+    if (!target->max_inline_data_set && src->max_inline_data_set) {
+        target->max_inline_data = src->max_inline_data;
+        target->max_inline_data_set = true;
+    }
+}
+
+static bool inline is_credit_message(const mca_btl_iwarp_recv_frag_t *frag)
+{
+    mca_btl_iwarp_control_header_t* chdr =
+        (mca_btl_iwarp_control_header_t *) to_base_frag(frag)->segment.seg_addr.pval;
+    return (MCA_BTL_TAG_IB == frag->hdr->tag) &&
+        (MCA_BTL_IWARP_CONTROL_CREDITS == chdr->type);
+}
+
+static bool inline is_cts_message(const mca_btl_iwarp_recv_frag_t *frag)
+{
+    mca_btl_iwarp_control_header_t* chdr =
+        (mca_btl_iwarp_control_header_t *) to_base_frag(frag)->segment.seg_addr.pval;
+    return (MCA_BTL_TAG_IB == frag->hdr->tag) &&
+        (MCA_BTL_IWARP_CONTROL_CTS == chdr->type);
+}
+
+static int32_t atoi_param(char *param, int32_t dflt)
+{
+    if (NULL == param || '\0' == param[0]) {
+        return dflt ? dflt : 1;
+    }
+
+    return atoi(param);
+}
+
+static void init_apm_port(mca_btl_iwarp_device_t *device, int port, uint16_t lid)
+{
+    int index;
+    struct mca_btl_iwarp_module_t *btl;
+    for(index = 0; index < device->btls; index++) {
+        btl = (mca_btl_iwarp_module_t *) opal_pointer_array_get_item(device->device_btls, index);
+        /* Ok, we already have btl for the fist port,
+         * second one will be used for APM */
+        btl->apm_port = port;
+        btl->port_info.apm_lid = lid + btl->src_path_bits;
+        mca_btl_iwarp_component.apm_ports++;
+        BTL_VERBOSE(("APM-PORT: Setting alternative port - %d, lid - %d"
+                    ,port ,lid));
+    }
+}
+
+static int get_var_source (const char *var_name, mca_base_var_source_t *source)
+{
+    int vari = mca_base_var_find ("opal", "btl", "iwarp", var_name);
+    if (0 > vari) {
+        return vari;
+    }
+
+    return mca_base_var_get_value (vari, NULL, source, NULL);
+}
+
+static int setup_qps(void)
+{
+    char **queues, **params = NULL;
+    int num_xrc_qps = 0, num_pp_qps = 0, num_srq_qps = 0, qp = 0;
+    uint32_t max_qp_size, max_size_needed;
+    int32_t min_freelist_size = 0;
+    int smallest_pp_qp = INT_MAX, ret = OPAL_ERROR;
+
+    queues = opal_argv_split(mca_btl_iwarp_component.receive_queues, ':');
+    if (0 == opal_argv_count(queues)) {
+        opal_show_help("help-mpi-btl-iwarp.txt",
+                       "no qps in receive_queues", true,
+                       opal_process_info.nodename,
+                       mca_btl_iwarp_component.receive_queues);
+        ret = OPAL_ERROR;
+        goto error;
+    }
+
+    while (queues[qp] != NULL) {
+        if (0 == strncmp("P,", queues[qp], 2)) {
+            num_pp_qps++;
+            if (smallest_pp_qp > qp) {
+                smallest_pp_qp = qp;
+            }
+        } else if (0 == strncmp("S,", queues[qp], 2)) {
```

Comment:
Are SRQs valid in iWARP?  If not, then don't even allow `S` in the receive queues specification.

And if that's true -- if there's only *one* kind of receive queue (point to point), then it might make the whole receive queues specification dramatically simpler.